opensfm/src/sfm/tracks_manager.h

```diff
@@ -0,0 +1,259 @@
+#pragma once
+
+#include <sfm/keypoint.h>
+#include <sfm/types.h>
+
+#include <fstream>
+#include <map>
+#include <unordered_map>
+#include <vector>
+
+class TracksManager {
+ public:
+  void AddTrack(const TrackId& id, const std::unordered_map<ShotId, Keypoint>& track) {
+    for (const auto observation : track) {
+      tracks_per_shot_[observation.first][id] = observation.second;
+      shot_per_tracks_[id][observation.first] = observation.second;
+    }
+  }
+
+  void AddObservation(const ShotId& shot_id, const TrackId& track_id, const Keypoint& observation) {
+    tracks_per_shot_[shot_id][track_id] = observation;
+    shot_per_tracks_[track_id][shot_id] = observation;
+  }
+
+  std::vector<ShotId> GetShotIds() const {
+    std::vector<ShotId> shots;
+    for (const auto& it : tracks_per_shot_) {
+      shots.push_back(it.first);
+    }
+    return shots;
+  }
+
+  std::vector<TrackId> GetTrackIds() const {
+    std::vector<TrackId> tracks;
```

Comment:
Reserve to avoid repeated memory allocation: tracks.reserve(shot_per_tracks_.size())